dsl_parser/functions.py

```diff
@@ -592,6 +547,103 @@ def _validate_ref(self, ref, ref_name):
                                        self.attribute_path))
 
 
+def _get_attribute_from_node_instance(ni, node, attribute_path, path):
+    value = _get_property_value(ni['node_id'],
+                                ni.get('runtime_properties'),
+                                attribute_path,
+                                path,
+                                raise_if_not_found=False)
+    if value is None:
+        # attribute not found in instance runtime properties
+        if value is None:
+            # special case for { get_attributes_list: [
+            #                        ..., node_instance_id] }
+            # returns the node-instance-id
+            if len(attribute_path) == 1\
+                    and attribute_path[0] == 'node_instance_id':
+                value = ni['id']
+    # still nothing? look in node properties
+    if value is None:
+        value = _get_property_value(node['id'],
+                                    node.get('properties', {}),
+                                    attribute_path,
+                                    path,
+                                    raise_if_not_found=False)
+    return value
+
+
+def _validate_no_functions_in_args(node_name, attribute_path, path, name,
+                                   scope, plan):
+    # If any of the arguments are functions don't validate
+    args = [node_name] + attribute_path
+    if any(is_function(x) for x in args):
+        return
+    if scope == scan.OUTPUTS_SCOPE and node_name in [SELF,
+                                                     SOURCE,
+                                                     TARGET]:
+        raise ValueError('{0} cannot be used with {1} function in '
+                         '{2}.'.format(node_name,
+                                       name,
+                                       path))
+    if scope == scan.NODE_TEMPLATE_SCOPE and \
+            node_name in [SOURCE, TARGET]:
+        raise ValueError('{0} cannot be used with {1} function in '
+                         '{2}.'.format(node_name,
+                                       name,
+                                       path))
+    if scope == scan.NODE_TEMPLATE_RELATIONSHIP_SCOPE and \
+            node_name == SELF:
+        raise ValueError('{0} cannot be used with {1} function in '
+                         '{2}.'.format(node_name,
+                                       name,
+                                       path))
```

Comment:
perhaps we can DRY this thing, because the very same error is raised 3 times in this function. A more involved `if`, to put all these conditions into a single `or`, perhaps?